samcli/lib/bootstrap/companion_stack/companion_stack_builder.py

```diff
@@ -0,0 +1,130 @@
+"""
+    Companion stack template builder
+"""
+import json
+
+from typing import Dict
+
+from samcli.lib.bootstrap.companion_stack.data_types import CompanionStack, ECRRepo
+from samcli import __version__ as VERSION
+
+
+class CompanionStackBuilder:
+    """
+    CFN template builder for the companion stack
+    """
+
+    _parent_stack_name: str
+    _companion_stack: CompanionStack
+    _repo_mapping: Dict[str, ECRRepo]
+
+    def __init__(self, companion_stack: CompanionStack) -> None:
+        self._companion_stack = companion_stack
+        self._repo_mapping: Dict[str, ECRRepo] = dict()
+
+    def add_function(self, function_logical_id: str) -> None:
+        """
+        Add an ECR repo associated with the function to the companion stack template
+        """
+        self._repo_mapping[function_logical_id] = ECRRepo(self._companion_stack, function_logical_id)
+
+    def clear_functions(self) -> None:
+        """
+        Remove all functions that need ECR repos
+        """
+        self._repo_mapping = dict()
+
+    def build(self) -> str:
+        """
+        Build companion stack CFN template with current functions
+        Returns
+        -------
+        str
+            CFN template for companions stack
+        """
+        template_dict = self._build_template_dict()
+        for _, ecr_repo in self._repo_mapping.items():
+            template_dict["Resources"][ecr_repo.logical_id] = self._build_repo_dict(ecr_repo)
+            template_dict["Outputs"][ecr_repo.output_logical_id] = CompanionStackBuilder._build_output_dict(ecr_repo)
+
+        return json.dumps(template_dict)
+
+    def _build_template_dict(self) -> Dict:
+        """
+        Build Companion stack template dictionary with Resources and Outputs not filled
+        Returns
+        -------
+        dict
+            Companion stack template dictionary
+        """
+        template = {
+            "AWSTemplateFormatVersion": "2010-09-09",
+            "Transform": "AWS::Serverless-2016-10-31",
```

Comment:
Technically you dont need a transform, since there are no SAM resources.

Reply:
I think this is only the base template, resources are added after at L46